AdvLoggerPkg/Library/AdvancedLoggerLib/MmCoreArm/AdvancedLoggerLib.c

```diff
@@ -102,48 +54,59 @@ AdvancedLoggerGetLoggerInfo (
   VOID
   )
 {
-  EFI_STATUS            Status;
-  EFI_PHYSICAL_ADDRESS  Address;
+  ADVANCED_LOGGER_INFO  *LoggerInfo;
+  EFI_PHYSICAL_ADDRESS  MaxAddress;
+  UINT32                HwPrintLevel;
 
   ASSERT (FeaturePcdGet (PcdAdvancedLoggerFixedInRAM));
   if (!FeaturePcdGet (PcdAdvancedLoggerFixedInRAM)) {
     return NULL;
   }
 
-  if (!mInitialized) {
-    // If this is the first time for MM core to get here, the memory attributes of this module
-    // may not be fully set yet. Thus set the memory for global variables attributes to RW first.
-    Address = ALIGN_VALUE ((EFI_PHYSICAL_ADDRESS)(UINTN)&mInitialized - EFI_PAGE_SIZE + 1, EFI_PAGE_SIZE);
-    Status  = MmuSetAttributes (Address, EFI_PAGE_SIZE, EFI_MEMORY_XP);
-    ASSERT_EFI_ERROR (Status);
-    Status = MmuClearAttributes (Address, EFI_PAGE_SIZE, EFI_MEMORY_RO);
-    ASSERT_EFI_ERROR (Status);
-
-    Address = ALIGN_VALUE ((EFI_PHYSICAL_ADDRESS)(UINTN)&mLoggerInfo - EFI_PAGE_SIZE + 1, EFI_PAGE_SIZE);
-    Status  = MmuSetAttributes (Address, EFI_PAGE_SIZE, EFI_MEMORY_XP);
-    ASSERT_EFI_ERROR (Status);
-    Status = MmuClearAttributes (Address, EFI_PAGE_SIZE, EFI_MEMORY_RO);
-    ASSERT_EFI_ERROR (Status);
-
-    mInitialized = TRUE;            // Only allow initialization once
-    mLoggerInfo  = (ADVANCED_LOGGER_INFO *)(VOID *)FixedPcdGet64 (PcdAdvancedLoggerBase);
-    ASSERT (mLoggerInfo != NULL);
-    if (mLoggerInfo == NULL) {
-      return NULL;
-    }
-
-    mLoggerInfo->HwPrintLevel = FixedPcdGet32 (PcdAdvancedLoggerHdwPortDebugPrintErrorLevel);
-
-    mMaxAddress = mLoggerInfo->LogBuffer + mLoggerInfo->LogBufferSize;
-    mBufferSize = mLoggerInfo->LogBufferSize;
+  LoggerInfo = (ADVANCED_LOGGER_INFO *)(VOID *)FixedPcdGet64 (PcdAdvancedLoggerBase);
+  if (LoggerInfo == NULL) {
+    return NULL;
+  }
+
+  // Initialize HdwPrintLevel if needed.
+  HwPrintLevel = FixedPcdGet32 (PcdAdvancedLoggerHdwPortDebugPrintErrorLevel);
+  if (LoggerInfo->HwPrintLevel != HwPrintLevel) {
+    LoggerInfo->HwPrintLevel = HwPrintLevel;
+  }
+
+  //
+  // The pointers LogBuffer and LogCurrent, and LogBufferSize, could be written
+  // to by untrusted code.  Here, we check that the pointers are within the
+  // allocated LoggerInfo space, and that LogBufferSize, which is used in
+  // multiple places to see if a new message will fit into the log buffer, is
+  // valid.
+  //
+
+  if (LoggerInfo->Signature != ADVANCED_LOGGER_SIGNATURE) {
+    return NULL;
   }
 
-  if (((mLoggerInfo) != NULL) && !ValidateInfoBlock ()) {
-    mLoggerInfo = NULL;
-    DEBUG ((DEBUG_ERROR, "%a: LoggerInfo marked invalid\n", __FUNCTION__));
+  // Ensure the start of the log is in the correct location.
+  if (LoggerInfo->LogBuffer != (PA_FROM_PTR (LoggerInfo + 1))) {
+    return NULL;
+  }
+
+  // Make sure the size of the buffer does not overrun it's fixed size.
+  if ((LoggerInfo->LogBuffer + LoggerInfo->LogBufferSize) >

```

Comment:
Is `LoggerInfo->LogBuffer` a pointer to an address? Should this just be comparing `LoggerInfo->LogBufferSize` against the size computed from the PCD?

Reply:
good catch